### PR TITLE
fix: check if the pattern upload, is valid

### PR DESCRIPTION
**Pull request by harkiratsm**
**Notes for Reviewers**

This PR fixes #

The designers sometimes get stuck when they try to upload docker-compose or a kubernetes manifest under design, since it is an invalid design construct.

Now we'll check for valid design files and only save them if they're valid, when users uploads.

**[Signed commits](https://github.com/meshery/meshery/blob/master/CONTRIBUTING.md#signing-off-on-commits-developer-certificate-of-origin)**
- [x] Yes, I signed my commits.

### CODE REVIEW

File: server/models/pattern/core/pattern.go [hunk 1]
```diff
@@ -207,6 +208,26 @@ func NewPatternFile(yml []byte) (af Pattern, err error) {
...
+// isValidPattern checks if the pattern file is valid or not
+func IsValidPattern(stringifiedFile string) (err error) {
+	pattern := Pattern{}
+
+	if err = yaml.Unmarshal([]byte(stringifiedFile), &pattern); err != nil {
+		return err
+	}
+
+	if pattern.Services == nil {
+		return errors.New("missing services field")
```
Comment (Abhishek-kumar09) [inline server/models/pattern/core/pattern.go:220]:
```suggestion
		return errors.New("invalid design-file format: missing services field")
```

File: server/models/pattern/core/pattern.go [hunk 2]
```diff
@@ -207,6 +208,26 @@ func NewPatternFile(yml []byte) (af Pattern, err error) {
...
+	}
+
+	for serviceName, service := range pattern.Services {
+		if service.Traits == nil {
+			return errors.New("missing traits field for:" + serviceName)
+		}
+	}
```
Comment (Abhishek-kumar09) [inline server/models/pattern/core/pattern.go:227]: Not required for a pattern/design to always have traits, although we have traits for every pattern/design now, this is not an important check.

**Review by Abhishek-kumar09 — CHANGES_REQUESTED** [review]

**Comment by leecalcote** [comment]
@harkiratsm did you address this feedback? If so, re-request a review.

**Comment by leecalcote** [comment]
and comment on how yo addressed the feedback.

File: server/models/pattern/core/pattern.go [hunk 3]
```diff
@@ -207,6 +208,26 @@ func NewPatternFile(yml []byte) (af Pattern, err error) {
...
+
+	if pattern.Services == nil {
+		return errors.New("invalid design-file format: missing services field")
+	}
+
+	for serviceName, service := range pattern.Services {
+		if service.Traits == nil {
+			return errors.New("missing traits field for:" + serviceName)
+		}
+	}
```
Comment (leecalcote) [inline server/models/pattern/core/pattern.go:227]:
```suggestion
//	for serviceName, service := range pattern.Services {
//		if service.Traits == nil {
//			return errors.New("missing traits field for:" + serviceName)
//		}
//	}
```

**Pull request merged**